### PR TITLE
Moving string to compiler_fn logic to eval_frame

### PR DESCRIPTION
**Pull request by anijain2305**
Helps with checking if a backend has any special context manager, like `aot_nvfuser` sets up torch.jit.fuser("fuser2") ctx mananger.

@jansel @pyjhzwh @ngimel

### CODE REVIEW

**Comment by ngimel** [comment]
Why would `torch.jit.fuser("fuser2") ` be needed if `--nvfuser` that's set as part of `--aot_nvfuser` is supposed to enabled it?

**Comment by anijain2305** [comment]
> Why would `torch.jit.fuser("fuser2") ` be needed if `--nvfuser` that's set as part of `--aot_nvfuser` is supposed to enabled it?

We are not setting `torch.jit.fuser("fuser2")` explicitly. TorchDynamo supports backend-specific context manager here - https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/eval_frame.py#L72. For aot_nvfuser, we have set the backend context manager here - https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/optimizations/training.py#L178


This was all working fine until we added a string. The string gets converted to a compiler_fn object later in the flow. So, I basically, pre-poned the conversion of string to the compiler_fn, so that we have access to the backend-specific constructor.

So, you would not need `--nvfuser` while calling benchmarking infra after this PR.

File: torchdynamo/eval_frame.py [hunk 1]
```diff
@@ -228,6 +242,9 @@ def toy_example(a, b):
         with torchdynamo.optimize(my_compiler):
            ...
     """
+    backend = get_compiler_fn(backend)
```
Comment (jansel) [inline torchdynamo/eval_frame.py:245]: Do we need this patch in optimize_assert too?

Comment (anijain2305) [reply torchdynamo/eval_frame.py:245]: Yes, its there further below.

**Review by jansel — APPROVED** [review]

**Pull request merged**